docs/src/modules/ROOT/pages/quickstart/school-timetabling/school-timetabling-constraints.adoc

```diff
@@ -122,5 +172,150 @@ public class TimetableConstraintProvider implements ConstraintProvider {
+            )
+            .penalize(HardSoftScore.ONE_HARD)
+            .justifyWith({ lesson1: Lesson, lesson2: Lesson?, score: HardSoftScore? ->
+                StudentGroupConflictJustification(lesson1.studentGroup, lesson1, lesson2!!)})
+            .asConstraint("Student group conflict");
+    }
+
+    fun teacherRoomStability(constraintFactory: ConstraintFactory): Constraint {
+        // A teacher prefers to teach in a single room.
+        return constraintFactory
+            .forEachUniquePair(
+                Lesson::class.java,
+                Joiners.equal(Lesson::teacher)
+            )
+            .filter { lesson1: Lesson, lesson2: Lesson -> lesson1.room !== lesson2.room }
+            .penalize(HardSoftScore.ONE_SOFT)
+            .justifyWith({ lesson1: Lesson, lesson2: Lesson?, score: HardSoftScore? ->
+                TeacherRoomStabilityJustification(lesson1.teacher, lesson1, lesson2!!)})
+            .asConstraint("Teacher room stability");
+    }
+
+    fun teacherTimeEfficiency(constraintFactory: ConstraintFactory): Constraint {
+        // A teacher prefers to teach sequential lessons and dislikes gaps between lessons.
+        return constraintFactory
+            .forEach(Lesson::class.java)
+            .join(Lesson::class.java,
+                Joiners.equal(Lesson::teacher),
+                Joiners.equal { lesson: Lesson -> lesson.timeslot?.dayOfWeek })
+            .filter { lesson1: Lesson, lesson2: Lesson ->
+                val between = Duration.between(
+                    lesson1.timeslot?.endTime,
+                    lesson2.timeslot?.startTime
+                )
+                !between.isNegative && between.compareTo(Duration.ofMinutes(30)) <= 0
+            }
+            .reward(HardSoftScore.ONE_SOFT)
+            .justifyWith({ lesson1: Lesson, lesson2: Lesson?, score: HardSoftScore? ->
+                TeacherTimeEfficiencyJustification(lesson1.teacher, lesson1, lesson2!!)})
+            .asConstraint("Teacher time efficiency");
+    }
+
+    fun studentGroupSubjectVariety(constraintFactory: ConstraintFactory): Constraint {
+        // A student group dislikes sequential lessons on the same subject.
+        return constraintFactory
+            .forEach(Lesson::class.java)
+            .join(Lesson::class.java,
+                Joiners.equal(Lesson::subject),
+                Joiners.equal(Lesson::studentGroup),
+                Joiners.equal { lesson: Lesson -> lesson.timeslot?.dayOfWeek })
+            .filter { lesson1: Lesson, lesson2: Lesson ->
+                val between = Duration.between(
+                    lesson1.timeslot?.endTime,
+                    lesson2.timeslot?.startTime
+                )
+                !between.isNegative && between.compareTo(Duration.ofMinutes(30)) <= 0
+            }
+            .penalize(HardSoftScore.ONE_SOFT)
+            .justifyWith({ lesson1: Lesson, lesson2: Lesson?, score: HardSoftScore? ->
```

Comment:
parentheses can be removed as the lambda is the only parameter.
all parameters are not-null in this case, aren't they? removing lesson2's "?" will eliminate the need for not-null casting ("!!") in the next line